toolchain/driver/clang_runner.cpp

```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
 namespace Carbon {
 
 ClangRunner::ClangRunner(const InstallPaths* install_paths,
-                         llvm::StringRef target,
                          llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs,
                          llvm::raw_ostream* vlog_stream)
     : ToolRunnerBase(install_paths, vlog_stream),
-      target_(target),
       fs_(std::move(fs)),
       diagnostic_ids_(new clang::DiagnosticIDs()) {}
 
+ClangRunner::~ClangRunner() {
+  if (!tmp_dir_.empty()) {
+    std::error_code ec;
+    if (auto number = std::filesystem::remove_all(tmp_dir_, ec); number == 0) {
+      CARBON_VLOG("Expected a temporary directory and found none: {0}",
+                  tmp_dir_);
+    } else if (ec) {
+      CARBON_VLOG("Error when cleaning temporary directory `{0}`: {1}",
+                  tmp_dir_, ec.message());
+    }
+  }
+}
+
+static auto ComputeTarget(llvm::ArrayRef<llvm::StringRef> args) -> std::string {
+  std::string target = llvm::sys::getDefaultTargetTriple();
+  for (auto [i, arg] : llvm::enumerate(args)) {
+    if (llvm::StringRef arg_copy = arg; arg_copy.consume_front("--target=")) {
+      target = arg_copy.str();
+    } else if ((arg == "--target" || arg == "-target") &&
+               (i + 1) < args.size()) {
+      target = args[i + 1].str();
+    } else if (arg == "--driver-mode=cl" ||
+               ((arg == "--driver-mode" || arg == "-driver-mode") &&
+                (i + 1) < args.size() && args[i + 1] == "cl")) {
+      // The `cl.exe` compatible driver mode should force to a
+      // `...-pc-windows-msvc` target.
+      llvm::Triple triple(target);
+      triple.setVendor(llvm::Triple::PC);
+      triple.setOS(llvm::Triple::Win32);
+      triple.setEnvironment(llvm::Triple::MSVC);
+      target = triple.str();
+    } else if (arg == "-m32") {
+      llvm::Triple triple(target);
+      if (!triple.isArch32Bit()) {
+        target = triple.get32BitArchVariant().str();
+      }
+    } else if (arg == "-m64") {
+      llvm::Triple triple(target);
+      if (!triple.isArch64Bit()) {
+        target = triple.get64BitArchVariant().str();
+      }
+    }
+  }
+  return target;
+}
+
+// Tries to detect a a non-linking list of Clang arguments to avoid setting up
+// the more complete resource directory needed for linking. False negatives are
+// fine here, and we use that to keep things simple.
+static auto IsNonLinkCommand(llvm::ArrayRef<llvm::StringRef> args) -> bool {
+  return llvm::any_of(args, [](llvm::StringRef arg) {
+    // Only check the most common cases as we have to do this for each argument.
+    if (arg.empty() || arg[0] != '-') {
+      return false;
+    }
+    if (arg.size() == 2) {
+      return arg[1] == 'c' || arg[1] == 'E' || arg[1] == 'S';
+    }
+    if (arg == "-fsyntax-only" || arg == "--version" || "--help" || "/?") {
```

Comment:
Won't `/?` hit the `!= '-'` test? Also note missing comparisons.

```suggestion
    if (arg == "-fsyntax-only" || arg == "--version" || arg == "--help") {
```

Reply:
Doh!

And really, I think I was overcomplicating all of this. Replaced it with something much less likely to get wrong. Sadly, its somewhat annoying to test this as it mostly just results in things being slow...